src/licenses/LicenseModel.cpp

```diff
@@ -0,0 +1,184 @@
+#include "LicenseModel.h"
+
+#include <QDirIterator>
+
+static const char *const platformRootPath = ":/licenses";
+static const char *const qtRootPath = ":/licenses/qt";
+static const char *const toolkitRootPath = ":/licenses/qt/licenses";
+static const char *const webviewRootPath = ":/licenses/qt/licenses/qtwebengine/src/3rdparty/chromium";
+
+namespace licenses {
+
+  LicenseModel::LicenseModel(QObject *parent) : QAbstractItemModel(parent), m_dir(platformRootPath) {
+    populate();
+  }
+
+  LicenseModel::~LicenseModel() = default;
+
+  void LicenseModel::populate() {
+    QList<QVariant> root_categories;
+    QVector<QVariant> rootData = {QString("Licenses"), QString(platformRootPath)};
+    rootItem = std::make_unique<LicenseItem>(rootData, root_categories);
+
+    QHash<QString, LicenseItem *> map;
+    map[rootItem->path()] = rootItem.get();
+
+    QDirIterator it(m_dir, QDirIterator::Subdirectories);
+
+    while (it.hasNext()) {
+      QString path = it.next();
+
+      QList<QVariant> categories;
+      categories.append(QVariant(static_cast<int>(LicenseCategory::All)));
+
+      if (path.startsWith(webviewRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::WebView)));
+
+      if (path.startsWith(toolkitRootPath) && !path.startsWith(webviewRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::Toolkit)));
+
+      if (path.startsWith(platformRootPath) && !path.startsWith(qtRootPath))
+        categories.append(QVariant(static_cast<int>(LicenseCategory::Platform)));
+
+      const QFileInfo &fileInfo = it.fileInfo();
+      LicenseItem *parent = map[fileInfo.dir().path()];
+
+      QVector<QVariant> child_data = {fileInfo.fileName(), fileInfo.absoluteFilePath()};
+      auto *child = new LicenseItem(child_data, categories, parent);
+      parent->appendChild(child);
+
+      if (fileInfo.isDir())
+        map[child->path()] = child;
+    }
+  }
+
+  QModelIndex LicenseModel::index(int row, int column, const QModelIndex &parent) const {
+    if (!hasIndex(row, column, parent))
+      return QModelIndex();
+
+    LicenseItem *parentItem;
+
+    if (!parent.isValid())
+      parentItem = rootItem.get();
+    else
+      parentItem = static_cast<LicenseItem *>(parent.internalPointer());
+
+    LicenseItem *childItem = parentItem->child(row);
+    if (childItem)
+      return createIndex(row, column, childItem);
+    return QModelIndex();
+  }
+
+  QModelIndex LicenseModel::parent(const QModelIndex &index) const {
```

Comment:
`parent` is both a function and also an often used local variable name.  Probably not a big deal, but something to think about.

Could call this `parentIndex`?

Reply:
Yes, I changed names, so now we have parentIndex (QModelIndex), parentItem (LicenseItem) and parentObject (QObject)